kyuubi-server/src/main/scala/org/apache/kyuubi/engine/reblance/RebalancedSparkEnginePolicy.scala

```diff
@@ -0,0 +1,190 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kyuubi.engine.reblance
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+import org.apache.commons.lang.StringUtils
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.config.KyuubiConf.{ENGINE_INIT_TIMEOUT, ENGINE_POOL_BALANCE_EXPAND_THRESHOLD, ENGINE_POOL_BALANCE_POLICY}
+import org.apache.kyuubi.ha.client.{DiscoveryClient, DiscoveryPaths}
+import org.apache.kyuubi.ha.client.DiscoveryClientProvider.withDiscoveryClient
+import org.apache.kyuubi.session.SessionHandle
+
+object RebalancedSparkEnginePolicy extends Logging {
+
+  private val sessionNodeSuffix = "-session"
+
+  def deleteHandlePath(
```

Comment:
"handle" is really a confusing name, we have "session handle", "engine handle", "operator handle", etc. and naming "handle" means nothing.